pkg/i2gw/providers/gce/gce_extensions.go

```diff
@@ -0,0 +1,201 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"context"
+	"encoding/json"
+
+	gkegatewayv1 "github.com/GoogleCloudPlatform/gke-gateway-api/apis/networking/v1"
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/ir"
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/notifications"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	backendconfigv1 "k8s.io/ingress-gce/pkg/apis/backendconfig/v1"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+type serviceNames []types.NamespacedName
+
+func getBackendConfigMapping(ctx context.Context, storage *storage) map[types.NamespacedName]serviceNames {
+	beConfigToSvcs := make(map[types.NamespacedName]serviceNames)
+
+	for _, service := range storage.Services {
+		svc := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+		ctx = context.WithValue(ctx, serviceKey, service)
+
+		// Read BackendConfig based on v1 BackendConfigKey.
+		beConfigName, exists := getBackendConfigName(ctx, service, backendConfigKey)
+		if exists {
+			beConfigKey := types.NamespacedName{Namespace: service.Namespace, Name: beConfigName}
+			beConfigToSvcs[beConfigKey] = append(beConfigToSvcs[beConfigKey], svc)
+			continue
+		}
+
+		// Read BackendConfig based on v1beta1 BackendConfigKey.
+		beConfigName, exists = getBackendConfigName(ctx, service, betaBackendConfigKey)
+		if exists {
+			beConfigKey := types.NamespacedName{Namespace: service.Namespace, Name: beConfigName}
+			beConfigToSvcs[beConfigKey] = append(beConfigToSvcs[beConfigKey], svc)
+			continue
+		}
+	}
+	return beConfigToSvcs
+}
+
+// Get names of the BackendConfig in the cluster based on the BackendConfig
+// annotation on k8s Services.
+func getBackendConfigName(ctx context.Context, service *apiv1.Service, backendConfigKey string) (string, bool) {
+	val, exists := getBackendConfigAnnotation(service, backendConfigKey)
+	if !exists {
+		return "", false
+	}
+
+	return parseBackendConfigName(ctx, val)
+}
+
+// Get the backend config annotation from the K8s service if it exists.
+func getBackendConfigAnnotation(service *apiv1.Service, backendConfigKey string) (string, bool) {
+	val, ok := service.Annotations[backendConfigKey]
+	if ok {
+		return val, ok
+	}
+	return "", false
+}
+
+type backendConfigs struct {
+	Default string            `json:"default,omitempty"`
+	Ports   map[string]string `json:"ports,omitempty"`
+}
+
+// Parse the name of the BackendConfig based on the annotation only if the same
+// BackendConfig is used for all Service ports.
```

Comment:
Nit: This comment is out of date now that you're picking a BackendConfig even if multiple ports are configured.

Reply:
Updated comment, thanks!